src/sas/qtgui/Perspectives/Fitting/ComplexConstraint.py

```diff
@@ -101,7 +102,7 @@ def setupParamWidgets(self):
         """
         self.cbParam1.clear()
         tab_index1 = self.cbModel1.currentIndex()
-        items1 = [param for param in self.params[tab_index1] if not self.tabs[tab_index1].paramHasConstraint(param)]
+        items1 = [param for param in self.tabs[tab_index1].main_params_to_fit]
```

Comment:
This will allow the same constraint to be used against two separate, varying parameters.
E.g.
`M1.scale = M2.scale` and `M1.scale = M3.scale` to be fitted at the same time.
Is this what we want? I just wonder what the optimization process would look like for this case.
I have explicitly removed this case but maybe there was some discussion which I missed that made it a requested feature?

Reply:
Actually, 4.2 does even worse. if you set the constraints as above, the fitting completes, but the first constraint is silently overwritten and the fitted parameters have `M1.scale` equal to `M3.scale` only. `M2.scale` varies independently.

Reply:
I don't think that we want this, because as pointed by Piotr it would be a source of confusion. For me, it should be clear that the constraint definition is slave_parameter = master_parameter, so the slave parameter will follow the master one, which is the one varied on the fit. Probably this needs to be made clear in the documentation, as I guess for many users (probably including Lionel) this is not clear. For example, from a mathematical point of view, a user could expect that the 2 definitions above imply M1.scale = M2. scale = M3. scale (and that it does not matter much which of the 3 is chosen as master). If I remember right, there was a ticket already on this. I don't think that it is easy to handle this situation as soon as the expressions are more complex than M1.param = M2.param = ... = MN.param, so I'd suggest to keep things simple for now and only allow a unique definition for a slave parameter.
Nico, did you change this behaviour on purpose or it was a mistake?

Reply:
Sorry folks I have been covered over and unable to keep up much but will try to look at all this soon as well as constraints are important for us also.  But to @gonzalezma point, yes there is a ticket about that (quite old I think).  If I understand the arguments (and need to look more carefully to be sure) I think you are making the same argument as @pkienzle made to me at the time. From the user perspective however (like foolish me) it looks obvious that in fact it shouldn't matter which is master.  I understand that it may be behind the scenes and certainly the last thing we need is for the cure to be worse than the disease :-)  I will try to find that ticket and link it here when I get a chance - it may be in sasmodels though.

Reply:
Yes indeed, having this wouldn't make sense and would most likeley cause an error in the optimization engine, but from what I understand each time a constraint is added, it is stored in the child of the parameter at row 0, so adding `M1.scale = M2.scale` and then adding `M1.scale = M3.scale` should overwrite the first constraint, just like in 4.2 (hence the warning informing the user he is overwriting an existing constraint)?  I removed this limitation as it was mentioned in #1490

Reply:
If I remember right, Paul K made the point that the computer = is not the mathematical =. Here doing A=B followed by A=C will make A=C (but not A=B and A=C). As Piotr noted, doing this in 4.2 will just overwrite the constraint definition and ignore the first definition. The only way to go beyond this would be to do a "syntactic" analysis of the constraints defined to find that we have A=B=C, but I think this is far from trivial as the user starts to write something a little bit more complex. I'm afraid that this could be the cause of some obscure errors (so the cure being worse than the disease), so personally I would avoid it and I'd try to think how things can be made clear in the interface and the documentation. But I let the real software experts, such as Piotr, Paul K and Nico decide what can be done or not.

Reply:
yep .. now that you say it that is what I remember as well (about computer vs mat = not being = 😄 ) so indeed if too complicated to address under the hood best not to, but anything in the interface that makes it harder for the poor user to do the wrong thing is certainly a welcome improvement.

Reply:
Ohhhh... so if I want to tie three parameters together in a situation where we will call A 'master', you mean I'm supposed to do B=A and then C=B???

Reply:
This should work. But personally I find clearer to select one master parameter, let's say A and define the constraints as B=A and C=A.

Reply:
What could be done to force the selection of one master parameter would be to exclude slave parameters from the combobox

Reply:
Yes, I think this makes sense.In the above example by @smk78, you shouldn't be able to constrain C to B, because B has already been constrained to another parameter. You should constrain C directly to A, and B should not be in the list of parameters available for constraining (the left combobox).